File: backend/app/Enums/ApiErrorCode.php
<?php

declare(strict_types=1);

namespace App\Enums;

/**
 * The machine-readable half of the error contract (docs/05-api-standards.md).
 *
 * Clients branch on these, never on the human-readable message, which is free to
 * change wording or be translated. Adding a case is backwards compatible; changing
 * or removing one is a breaking API change and needs a new API version.
 */
enum ApiErrorCode: string
{
    case ValidationFailed = 'VALIDATION_FAILED';
    case Unauthenticated = 'UNAUTHENTICATED';
    case Forbidden = 'FORBIDDEN';
    case NotFound = 'NOT_FOUND';
    case MethodNotAllowed = 'METHOD_NOT_ALLOWED';
    case Conflict = 'CONFLICT';
    case IdempotencyKeyReused = 'IDEMPOTENCY_KEY_REUSED';
    case RateLimited = 'RATE_LIMITED';
    case BusinessRuleViolated = 'BUSINESS_RULE_VIOLATED';
    case DependencyUnavailable = 'DEPENDENCY_UNAVAILABLE';
    case ServerError = 'SERVER_ERROR';

    // --- Module 03: customer authentication -----------------------------
    case InvalidPhone = 'INVALID_PHONE';
    case UnsupportedPhoneRegion = 'UNSUPPORTED_PHONE_REGION';
    case OtpSendFailed = 'OTP_SEND_FAILED';
    case OtpRateLimited = 'OTP_RATE_LIMITED';
    case OtpInvalid = 'OTP_INVALID';
    case OtpExpired = 'OTP_EXPIRED';
    case OtpTooManyAttempts = 'OTP_TOO_MANY_ATTEMPTS';
    case OtpResendTooSoon = 'OTP_RESEND_TOO_SOON';
    case RegistrationTokenInvalid = 'REGISTRATION_TOKEN_INVALID';
    case RegistrationTokenExpired = 'REGISTRATION_TOKEN_EXPIRED';
    case AccountSuspended = 'ACCOUNT_SUSPENDED';
    case AccountDisabled = 'ACCOUNT_DISABLED';

    // --- Module 04: profile and saved addresses -------------------------
    case AddressLimitReached = 'ADDRESS_LIMIT_REACHED';
    case AddressNotFound = 'ADDRESS_NOT_FOUND';

    // Module 05
    case TripNotFound = 'TRIP_NOT_FOUND';
    case TripLimitReached = 'TRIP_LIMIT_REACHED';
    case TripNotEditable = 'TRIP_NOT_EDITABLE';

    public function httpStatus(): int
    {
        return match ($this) {
            self::ValidationFailed => 422,
            self::Unauthenticated => 401,
            self::Forbidden => 403,
            self::NotFound => 404,
            self::MethodNotAllowed => 405,
            self::Conflict, self::IdempotencyKeyReused => 409,
            // 422, not 409: the request is well formed and the conflict is with a
            // limit rather than with another version of the same resource.
            self::AddressLimitReached => 422,
            // 404, and deliberately the same answer an address that does not
            // exist gets — see CustomerAddressService::ownedByOrFail().
            self::AddressNotFound => 404,
            // Same reasoning again, for journeys: not-yours and does-not-exist
            // are one answer, so the endpoint cannot be walked to discover which
            // identifiers are real.
            self::TripNotFound => 404,
            self::TripLimitReached,
            // 422 rather than 409: the request is valid, the journey is simply
            // past the point where changing it means anything.
            self::TripNotEditable => 422,
            self::RateLimited => 429,
            self::BusinessRuleViolated => 422,
            self::DependencyUnavailable => 503,
            self::ServerError => 500,

            self::InvalidPhone,
            self::UnsupportedPhoneRegion,
            self::OtpInvalid,
            self::OtpExpired,
            self::OtpTooManyAttempts => 422,

            // 429 for both: a resend asked for too early is a rate limit, and
            // giving it its own status would let a caller distinguish "too soon"
            // from "too many" and tune an abuse loop against it.
            self::OtpRateLimited,
            self::OtpResendTooSoon => 429,

            self::RegistrationTokenInvalid,
            self::RegistrationTokenExpired => 401,

            // 403, not 401: the caller proved who they are. They are not allowed.
            self::AccountSuspended,
            self::AccountDisabled => 403,

            self::OtpSendFailed => 503,
        };
    }
}

File: mobile/lib/core/network/api_error_code.dart
/// The machine-readable error codes the API can return.
///
/// These mirror `App\Enums\ApiErrorCode` on the backend. The app branches on
/// these and **never** on the human message: the message is written for a
/// person, may be translated, and is expected to change. A client that parses
/// prose breaks the first time somebody improves the wording.
///
/// [unknown] is not a failure of this list — a newer server may send a code an
/// older app has never heard of, and the app must degrade to a generic message
/// rather than crash.
enum ApiErrorCode {
  validationFailed('VALIDATION_FAILED'),
  unauthenticated('UNAUTHENTICATED'),
  forbidden('FORBIDDEN'),
  notFound('NOT_FOUND'),
  methodNotAllowed('METHOD_NOT_ALLOWED'),
  conflict('CONFLICT'),
  idempotencyKeyReused('IDEMPOTENCY_KEY_REUSED'),
  rateLimited('RATE_LIMITED'),
  businessRuleViolated('BUSINESS_RULE_VIOLATED'),
  dependencyUnavailable('DEPENDENCY_UNAVAILABLE'),
  serverError('SERVER_ERROR'),

  invalidPhone('INVALID_PHONE'),
  unsupportedPhoneRegion('UNSUPPORTED_PHONE_REGION'),
  otpSendFailed('OTP_SEND_FAILED'),
  otpRateLimited('OTP_RATE_LIMITED'),
  otpInvalid('OTP_INVALID'),
  otpExpired('OTP_EXPIRED'),
  otpTooManyAttempts('OTP_TOO_MANY_ATTEMPTS'),
  otpResendTooSoon('OTP_RESEND_TOO_SOON'),
  registrationTokenInvalid('REGISTRATION_TOKEN_INVALID'),
  registrationTokenExpired('REGISTRATION_TOKEN_EXPIRED'),
  accountSuspended('ACCOUNT_SUSPENDED'),
  accountDisabled('ACCOUNT_DISABLED'),

  addressLimitReached('ADDRESS_LIMIT_REACHED'),
  addressNotFound('ADDRESS_NOT_FOUND'),

  tripNotFound('TRIP_NOT_FOUND'),
  tripLimitReached('TRIP_LIMIT_REACHED'),

  /// The journey has departed or been cancelled, so there is nothing left to
  /// change. Almost always means the screen in front of the customer is stale.
  tripNotEditable('TRIP_NOT_EDITABLE'),

  /// The request never reached the server, or never came back.
  network('NETWORK'),

  /// A code this build does not know, or a response that was not the documented
  /// envelope at all.
  unknown('UNKNOWN');

  const ApiErrorCode(this.wire);

  /// The string the API actually sends.
  final String wire;

  static ApiErrorCode fromWire(String? value) {
    if (value == null) return ApiErrorCode.unknown;

    for (final ApiErrorCode code in ApiErrorCode.values) {
      if (code.wire == value) return code;
    }
    return ApiErrorCode.unknown;
  }

  /// Whether retrying the same request unchanged could plausibly succeed.
  ///
  /// Drives whether a screen offers "Try again" or asks the customer to change
  /// something first.
  bool get isRetryable => switch (this) {
    ApiErrorCode.network ||
    ApiErrorCode.serverError ||
    ApiErrorCode.dependencyUnavailable ||
    ApiErrorCode.otpSendFailed => true,
    _ => false,
  };
}
